[WIP] Support module in errata sync

dparalen opened the pull request:
This patch introduces:
* parsing of the new errata collection element:
  <module name="n" stream="s" version="v" context="c" arch="a" />

Fixes: 3919
https://pulp.plan.io/issues/3919

pep8speaks commented:
Hello @dparalen! Thanks for submitting the PR.



 - In the file [`plugins/test/unit/plugins/importers/yum/repomd/test_updateinfo.py`](https://github.com/pulp/pulp_rpm/blob/a7836b8d01a0b63066b37bade237289983969002/plugins/test/unit/plugins/importers/yum/repomd/test_updateinfo.py), following are the PEP8 issues :

> [Line 150:101](https://github.com/pulp/pulp_rpm/blob/a7836b8d01a0b63066b37bade237289983969002/plugins/test/unit/plugins/importers/yum/repomd/test_updateinfo.py#L150): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 100 characters)
> [Line 228:101](https://github.com/pulp/pulp_rpm/blob/a7836b8d01a0b63066b37bade237289983969002/plugins/test/unit/plugins/importers/yum/repomd/test_updateinfo.py#L228): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 100 characters)
> [Line 234:101](https://github.com/pulp/pulp_rpm/blob/a7836b8d01a0b63066b37bade237289983969002/plugins/test/unit/plugins/importers/yum/repomd/test_updateinfo.py#L234): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 100 characters)
> [Line 262:23](https://github.com/pulp/pulp_rpm/blob/a7836b8d01a0b63066b37bade237289983969002/plugins/test/unit/plugins/importers/yum/repomd/test_updateinfo.py#L262): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 274:23](https://github.com/pulp/pulp_rpm/blob/a7836b8d01a0b63066b37bade237289983969002/plugins/test/unit/plugins/importers/yum/repomd/test_updateinfo.py#L274): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

dparalen commented:
Just updated the patch with first take on the publish  update, could use some feedback:

```xml
[vagrant@pulp2 pulp_rpm]$ cat updateinfo.xml
<?xml version="1.0" encoding="utf-8"?>
<updates>
  <update status="stable" from="errata@redhat.com" version="1" type="enhancement">
    <id>RHEA-2012:0059</id>
    <issued date="2018-01-27 16:08:09" />
    <title>Duck_Kangaroo_Erratum</title>
    <release>1</release>
    <description>Duck_Kangaro_Erratum description</description>
    <updated date="2018-07-20 06:00:01 UTC" />
    <references />
    <pkglist>
      <collection short="modulerr_1">
        <name>modulerr_1</name>
        <module stream="0" version="20180730223407" arch="noarch" name="kangaroo" context="deadbeef" />
        <package src="http://www.fedoraproject.org" name="kangaroo" epoch="0" version="0.3" release="1" arch="noarch">
          <filename>kangaroo-0.3-1.noarch.rpm</filename>
        </package>
      </collection>
      <collection short="modulerr_2">
        <name>modulerr_2</name>
        <module stream="0" version="20180730233102" arch="noarch" name="duck" context="deadbeef" />
        <package src="http://www.fedoraproject.org" name="duck" epoch="0" version="0.7" release="1" arch="noarch">
          <filename>duck-0.7-1.noarch.rpm</filename>
        </package>
      </collection>
    </pkglist>
  </update>
  <update status="stable" from="errata@redhat.com" version="1" type="security">
    <id>RHEA-2012:0055</id>
    <issued date="2012-01-27 16:08:06" />
    <title>Sea_Erratum</title>
    <release>1</release>
    <description>Sea_Erratum</description>
    <updated date="2012-01-27 16:08:06" />
    <references />
    <pkglist>
      <collection short="modulerr_0">
        <name>modulerr_0</name>
        <package src="http://www.fedoraproject.org" name="walrus" epoch="0" version="5.21" release="1" arch="noarch">
          <filename>walrus-5.21-1.noarch.rpm</filename>
        </package>
        <package src="http://www.fedoraproject.org" name="penguin" epoch="0" version="0.9.1" release="1" arch="noarch">
          <filename>penguin-0.9.1-1.noarch.rpm</filename>
        </package>
        <package src="http://www.fedoraproject.org" name="shark" epoch="0" version="0.1" release="1" arch="noarch">
          <filename>shark-0.1-1.noarch.rpm</filename>
        </package>
      </collection>
    </pkglist>
  </update>
  <update status="stable" from="errata@redhat.com" version="1" type="security">
    <id>RHEA-2012:0056</id>
    <issued date="2013-01-27 16:08:08" />
    <title>Bird_Erratum</title>
    <release>1</release>
    <description>ParthaBird_Erratum</description>
    <updated date="2013-02-27 17:00:00" />
    <references />
    <pkglist>
      <collection short="modulerr_0">
        <name>modulerr_0</name>
        <package src="http://www.fedoraproject.org" name="crow" epoch="0" version="0.8" release="1" arch="noarch">
          <filename>crow-0.8-1.noarch.rpm</filename>
        </package>
        <package src="http://www.fedoraproject.org" name="stork" epoch="0" version="0.12" release="2" arch="noarch">
          <filename>stork-0.12-2.noarch.rpm</filename>
        </package>
        <package src="http://www.fedoraproject.org" name="duck" epoch="0" version="0.6" release="1" arch="noarch">
          <filename>duck-0.6-1.noarch.rpm</filename>
        </package>
      </collection>
    </pkglist>
  </update>
  <update status="stable" from="errata@redhat.com" version="1" type="security">
    <id>RHEA-2012:0057</id>
    <issued date="2013-01-27 16:08:05" />
    <title>Bear_ErratumPARTHA</title>
    <release>1</release>
    <description>Bear_Erratum</description>
    <updated date="2013-01-27 16:08:05 UTC" />
    <references />
    <pkglist>
      <collection short="modulerr_0">
        <name>modulerr_0</name>
        <package src="http://www.fedoraproject.org" name="bear" epoch="0" version="4.1" release="1" arch="noarch">
          <filename>bear-4.1-1.noarch.rpm</filename>
        </package>
      </collection>
    </pkglist>
  </update>
  <update status="stable" from="errata@redhat.com" version="1" type="enhancement">
    <id>RHEA-2012:0060</id>
    <issued date="2018-01-29 16:08:09" />
    <title>Duck_0.8_Erratum</title>
    <release>1</release>
    <description>Duck_0.8_Erratum description</description>
    <updated date="2018-07-29 06:00:01 UTC" />
    <references />
    <pkglist>
      <collection short="modulerr_1">
        <name>modulerr_1</name>
        <module stream="0" version="201809302113907" arch="noarch" name="duck" context="deadbeef" />
        <package src="http://www.fedoraproject.org" name="duck" epoch="0" version="0.8" release="1" arch="noarch">
          <filename>duck-0.8-1.noarch.rpm</filename>
        </package>
      </collection>
    </pkglist>
  </update>
  <update status="stable" from="errata@redhat.com" version="1" type="enhancement">
    <id>RHEA-2012:0058</id>
    <issued date="2013-01-27 16:08:09" />
    <title>Gorilla_Erratum</title>
    <release>1</release>
    <description>Gorilla_Erratum</description>
    <updated date="2014-07-20 06:00:01 UTC" />
    <references />
    <pkglist>
      <collection short="modulerr_0">
        <name>modulerr_0</name>
        <package src="http://www.fedoraproject.org" name="gorilla" epoch="0" version="0.62" release="1" arch="noarch">
          <filename>gorilla-0.62-1.noarch.rpm</filename>
        </package>
      </collection>
    </pkglist>
  </update>
</updates>
```

goosemania commented on plugins/pulp_rpm/plugins/distributors/yum/publish.py:
```diff
@@ -700,25 +700,58 @@ def _get_pkglist_to_publish(self, erratum_unit):
-
-        return new_pkglist
+                # ret[0] is special; all non-modular collections are aggregated there
+                new_pkglist = ret[0]
```
I'm not sure I see how all non-modular collections get aggregated in this collection, unless all the non-modular collections are present before any modular ones. Can you elaborate?

dparalen replied:
imagine the list of collections started with a modular collection; the first item in the `ret` list is an empty list of packages; `pkglist_index` gets bumped to `1` and a new, modular package list is created in `new_pkglist` and appended to the `ret` list, which now contains an empty list as the 0th item and the modular collection list of packages as the second item. After that, the `new_pkglist` points to the 0th element of `ret` again and the outermost loop repeats.

The invariant could be stated as:
* at any iteration, the 0th `ret` item is a list of all the non-modular collection packages encountered, accumulated
* every modular collection has a custom list, following the 0th `ret` item
* the `pkglist_index` is never decreased

Btw the 0th `ret` item can happen to have `pkglist_index > 1` which I hope is OK. After the naming update, the index name will be`collection_index`

dparalen replied:
actually I'm wrong here, the `pkglist_index` of the aggregated collection will always be 0, but might not be present in the erratum (I hope I've got it right this time ;)

goosemania commented on plugins/pulp_rpm/plugins/importers/yum/repomd/updateinfo.py:
```diff
@@ -106,6 +106,13 @@ def _parse_collection(element):
     }
+
+    # a module element indicates this erratum being part of a module
+    modulemds = element.findall('module')
```
Why modulemds are in plural? There could be 0 or 1 modulemd in a collection.

dparalen replied:
to match the return type of the `element.findall()` method

goosemania commented on plugins/pulp_rpm/plugins/importers/yum/repomd/updateinfo.py:
```diff
@@ -106,6 +106,13 @@ def _parse_collection(element):
+    modulemds = element.findall('module')
+    # a last-item-override logic applies
+    if modulemds:
+        ret['module'] = _parse_module(modulemds[-1])
```
I suggest to fail the import of an erratum which contains multiple modules in a collection.
Otherwise it's very hard to notice this error in an erratum format + there is no guarantee that the last element is a correct one.

dparalen replied:
OK, we should probably update our XML "schema" specification to explicitly state this limitation

goosemania commented on plugins/pulp_rpm/plugins/distributors/yum/publish.py:
```diff
@@ -662,34 +666,30 @@ def _get_repo_unit_nevra(self, repo_id):
             # human-readable string, like "A Human-Readable Repository Description". Pulp doesn't
             # require storing that sort of string with a repo, so repo_id is the best we've got.
-            'name': repo_id,
+            'name': '{}_{}'.format(repo_id, index),
```
repo_id + index as a name looks good to me. I think it's good enough and including a module name might not be necessary. At the same time it can be helpful for testing/troubleshooting purposes to have a module name as well, not essential though.

dparalen replied:
OK, would templates like`foo_repo_007_default` vs `bar_repo_42_betelgeuse` fit the non-modular vs modular collection naming respectively?

dparalen commented on plugins/pulp_rpm/plugins/distributors/yum/publish.py:
```diff
@@ -662,34 +666,30 @@ def _get_repo_unit_nevra(self, repo_id):
+    def _new_pkglist(self, repo_id, index):
+        # start a new pkglist/collection (for a module), ammend the pkglist name, record
+        # the module and add the pkglist into the returned list
+        proto_pkglist = {
```
cool, will update

dparalen commented on plugins/pulp_rpm/plugins/distributors/yum/publish.py:
```diff
@@ -700,25 +700,58 @@ def _get_pkglist_to_publish(self, erratum_unit):
+                        # not from the repo being published
+                        continue
+                    pkglist_index += 1
+                    new_pkglist = self._new_pkglist(repo_id, pkglist_index)
```
agreed

dparalen commented:
Hey @goosemania! 

Thanks for the review; please let me know whether I've addressed your concerns and I shall update the PR accordingly.

Cheers,
milan

dparalen commented:
Just updated with a commit that implements recursive copy

goosemania commented on plugins/pulp_rpm/plugins/distributors/yum/publish.py:
```diff
@@ -662,63 +666,90 @@ def _get_repo_unit_nevra(self, repo_id):
             # no pkglist will ever contain packages, so short out and return the empty pkglist
-            return new_pkglist
-
+            # Please note if there's no packages at all, no modules can be associated with a
```
I suggest to remove that comment or to rephrase it.
There are valid modules with no rpms in repositories.
In theory it's possible that there are only modules with no rpms and their metadata got updated.
I'm not suggesting to support this case, there is no practical use. But stating that no modules can be associated with a repo because of no rpms doesn't look correct to me. What do you think?

dparalen replied:
> There are valid modules with no rpms in repositories.

ah, didn't know this; I was pretty sure a repo without rpms wouldn't have any modules either, that might need updating by an erratum.  Unsure the logic here is sufficient/sound then.
What would you prefer me to do?

goosemania commented on plugins/pulp_rpm/plugins/distributors/yum/metadata/updateinfo.py:
```diff
@@ -105,9 +105,9 @@ def add_unit_metadata(self, item, filtered_pkglist):
-        :param filtered_pkglist: The pkglist containing unique non-empty collections
-                                with packages which are present in repo.
-        :type filtered_pkglist: dict
+        :param filtered_pkglist: The list of non-empty collections with packages which are present
```
I can see now that collections can be empty, even the special non-modular one. I suggest to `s/non-empty//`

dparalen replied:
will update; insufficient copy-paste handling

goosemania commented on plugins/pulp_rpm/plugins/distributors/yum/publish.py:
```diff
@@ -662,63 +666,90 @@ def _get_repo_unit_nevra(self, repo_id):
-        return new_pkglist
+                            new_collection['packages'].append(package)
+                # ret[0] is special; all non-modular package collections are aggregated there
+                new_collection = ret[0]
```
Thanks! Logic looks right this time :)

dparalen replied:
heh :)

goosemania commented on plugins/pulp_rpm/plugins/importers/yum/associate.py:
```diff
@@ -102,6 +103,17 @@ def associate(source_repo, dest_repo, import_conduit, config, units=None, solver
+    module_search_dicts = None
+    associated_units |= source_repo_modules - dest_repo_modules
+    source_repo_modules = None
+    dest_repo_modules = None
```
Why do you need to set those to None?

dparalen replied:
for an "immediate" garbage collection (optimization); see the comments around similar statements above

goosemania commented on plugins/pulp_rpm/plugins/db/models.py:
```diff
@@ -1190,6 +1198,26 @@ def get_unique_pkglists(cls, errata_id):
         return pkglists
 
+    @classmethod
+    def get_unique_modules(cls, errata_id):
```
Why should it be a classmethod?

dparalen replied:
copy-paste design of `get_unique_pkglists` but cool to update

goosemania commented on plugins/pulp_rpm/plugins/db/models.py:
```diff
@@ -1190,6 +1198,26 @@ def get_unique_pkglists(cls, errata_id):
+        match_stage = {'$match': {'errata_id': errata_id, 'collections.module': {'$exists': True}}}
+        project_stage = {'$project': {'collections.module': True}}
+        unwind_stage = {'$unwind': '$collections'}
+        group_stage = {'$group': {'_id': '$errata_id',
```
Can you elaborate why it's not enough to have just a match stage and a group stage? I probably miss something, mongo data manipulation is often counter-intuitive :)

dparalen replied:
`match` reduces the units processed and `project` minimizes the data amount processed. Just the `module` data needs to be picked. `unwind` multiplies the amount of units by popping them out of the nested lists --- this is cool thanks to the `project` stage. As a result, the `group` stage returns a flat structure. Can rework the same as the `pkglists` equivalent tho.

dparalen commented on docs/user-guide/release-notes/2.18.x.rst:
```diff
@@ -0,0 +1,11 @@
+New Features
+------------
+
+* Errata can now refer to modular package collections.
```
ack, will update

goosemania commented:
@dparalen , please, check the docs builder failure as well, you need to add new release notes file to the toc_tree I think

dparalen commented:
So, I ran some syntetic test case (since there are too little too short modular erratas out there, in the wild ;) to figure out possible performance issues with my patch w/r the `project` stage of my module search dicts aggregation pipeline. My setup is as follows.

* I set up pulp2 in dev environment, and synced the [modular fixture repo](https://repos.fedorapeople.org/pulp/pulp/fixtures/rpm-with-modules/).
* I then followed my custom modular [erratum upload gist](https://gist.github.com/dparalen/779f8018af1d7e090bb380a34acce851) to create a custom erratum that I could extend a lot to stress-test the aggregation pipeline
* having uploaded the erratum, I dropped it from manually from Mongo to prevent ID collisions:
 `db.erratum_pkglists.deleteOne({"errata_id": "RHEA-2018:0042"})`
* then I created a script like this one to create the massive erratum:
```python
import pymongo
c = MongoClient() 
db = c.pulp_database

def mk_item(collections):                                                                                                      
    return {
         'errata_id': 'RHEA-2018:0042',
         '_ns': 'erratum_pkglists',
         'repo_id': 'modulerr',
         'collections': collections,
   } 

def mk_collection(idx, packages):                                                                                              
     return {
         'module': {                   
                 'arch': 'noarch',     
                  'context': 'c0ffee42',
                  'name': 'duck',        
                  'stream': '0',                                                                                                                                                     
                  'version': '20180730233102',
             },                                                                                                                                                                     
             'packages': packages,
              'short': 'duck_{}'.format(idx),
              'name':  'duck_{}'.format(idx),
        }

def mk_pkg(idx):                                                                                                                                                              
         return {
              'arch': 'noarch',             
              'epoch': None,            
              'filename': 'duck-0.7.{}.noarch.rpm'.format(idx),
              'name': 'duck',            
              'release': '{}'.format(idx),                                                                                                                                          
              'src': 'http://www.fedoraproject.org',
              'sum': None,                                                                                                                                                          
              'version': '0.7',    
         }

# generate!!!
db.erratum_pkglists.insert_one(mk_item([mk_collection(i, [mk_pkg(j) for j in xrange(128*k, 128*(k+1))]) for i in xrange(128)]))
```
* I double checked there indeed were 32k items in the erratum in Mongo: `b.erratum_pkglists.find({"errata_id": "RHEA-2018:0042"})[0]`
```json
                                {
                                        "src" : "http://www.fedoraproject.org",
                                        "epoch" : null,
                                        "version" : "0.7",
                                        "name" : "duck",
                                        "release" : "32767",
                                        "sum" : null,
                                        "arch" : "noarch",
                                        "filename" : "duck-0.7.32767.noarch.rpm"
                                }
                        ],
                        "name" : "duck_127",
                        "module" : {
                                "stream" : "0",
                                "version" : "20180730233102",
                                "arch" : "noarch",
                                "name" : "duck",
                                "context" : "c0ffee42"
                        },
                        "short" : "duck_127"
                }
        ]
}
```
* then I created an empty repo `modulerr_copy` to recursively copy the erratum into and ran the copying couple of times to get some numbers:


```
[vagrant@pulp2 pulp_rpm]$ time pulp-admin rpm repo copy errata -f modulerr -t modulerr_copy --str-eq 'errata_id=RHEA-2018:0042' --recursive                                             
This command may be exited via ctrl+c without affecting the request.


[/]
Running...

Copied:
  RHEA-2018:0042


real    0m7.981s
user    0m0.572s
sys     0m0.092s
[vagrant@pulp2 pulp_rpm]$ time pulp-admin rpm repo copy errata -f modulerr -t modulerr_copy --str-eq 'errata_id=RHEA-2018:0042' --recursive                                             
This command may be exited via ctrl+c without affecting the request.


[/]
Running...

Copied:
  RHEA-2018:0042


real    0m7.904s
user    0m0.568s
sys     0m0.079s
[vagrant@pulp2 pulp_rpm]$ time pulp-admin rpm repo copy errata -f modulerr -t modulerr_copy --str-eq 'errata_id=RHEA-2018:0042' --recursive                                             
This command may be exited via ctrl+c without affecting the request.


[-]
Running...

Copied:
  RHEA-2018:0042


real    0m9.028s
user    0m0.599s
sys     0m0.117s
[vagrant@pulp2 pulp_rpm]$
```

* then I changed the code on the `Erratum` model to read:
```diff
diff --git a/plugins/pulp_rpm/plugins/db/models.py b/plugins/pulp_rpm/plugins/db/models.py
index db997d50..112d5a4b 100644
--- a/plugins/pulp_rpm/plugins/db/models.py
+++ b/plugins/pulp_rpm/plugins/db/models.py
@@ -1209,14 +1209,13 @@ class Errata(UnitMixin, ContentUnit):
         :rtype: a lists of dicts
         """
         match_stage = {'$match': {'errata_id': errata_id, 'collections.module': {'$exists': True}}}
-        project_stage = {'$project': {'collections.module': True}}
-        unwind_stage = {'$unwind': '$collections'}
         group_stage = {'$group': {'_id': '$errata_id',
                                   'modules': {'$addToSet': '$collections.module'}}}
-        modules = ErratumPkglist.objects.aggregate(match_stage, project_stage,
-                                                   unwind_stage, group_stage,
-                                                   allowDiskUse=True).next()['modules']
-        return modules
+        modules_collection = ErratumPkglist.objects.aggregate(
+            match_stage, group_stage, allowDiskUse=True).next()['modules']
+        for collection in modules_collection:
+            for module in collection:
+                yield module
 
     def create_legacy_metadata_dict(self):
         """

```
* restarted the services and checked again:

```[vagrant@pulp2 pulp_rpm]$ time pulp-admin rpm repo copy errata -f modulerr -t modulerr_copy --str-eq 'errata_id=RHEA-2018:0042' --recursive
This command may be exited via ctrl+c without affecting the request.


[/]
Running...

Copied:
  RHEA-2018:0042


real    0m7.967s
user    0m0.553s
sys     0m0.096s
[vagrant@pulp2 pulp_rpm]$ time pulp-admin rpm repo copy errata -f modulerr -t modulerr_copy --str-eq 'errata_id=RHEA-2018:0042' --recursive
This command may be exited via ctrl+c without affecting the request.


[-]
Running...

Copied:
  RHEA-2018:0042


real    0m8.962s
user    0m0.572s
sys     0m0.089s
[vagrant@pulp2 pulp_rpm]$ time pulp-admin rpm repo copy errata -f modulerr -t modulerr_copy --str-eq 'errata_id=RHEA-2018:0042' --recursive
This command may be exited via ctrl+c without affecting the request.


[/]
Running...

Copied:
  RHEA-2018:0042


real    0m7.915s
user    0m0.546s
sys     0m0.098s
[vagrant@pulp2 pulp_rpm]$ 
```

In conclusion, on my sample mega erratum, there's no difference in performance on the python side.
I guess we can drop the `project` and `unwind` stages for consistency sake.
